### PR TITLE
[WIP] Move the file config souce to it's own class. It's now extensible

### PR DESCRIPTION
**Pull request by MadsNielsen**
…e for #34 

WIP still im thinking of what the 'entry point' for this should be. Currently we expect no parameters, and no GUI but im open to ideas.

### CODE REVIEW

File: src/main/java/org/jenkinsci/plugins/casc/configurationsource/ConfigurationSource.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+import java.util.List;
+import java.util.Map;
+
+/**
+ * SPI For configurationsource source. All our sources are .yaml|.yml files, it's the method to which we resolve the value and
+ * where those values are stored we want to be able to configure.
+ */
+public abstract class ConfigurationSource implements ExtensionPoint {
+
+    public abstract Map<String, InputStream> listSources() throws IOException;
```
Comment (lanwen) [inline src/main/java/org/jenkinsci/plugins/casc/configurationsource/ConfigurationSource.java]: No way! Please don't make InputStream as public api in 2018. Its not safe at all! 
- Where should it be closed? 
- How to make it NIO-compatible? 
- What if I have ready-to-use objects? Why should I bother with InputStream?

Please please don't!

Also see #119

**Comment by lanwen** [comment]
Please move YAML part to extension point too. What if someday config will be in Kotlin? Or SnakeYaml will be replaced by Jackson?

**Comment by lanwen** [comment]
I'd like to suggest closest to plugin form of the raw data - `Map<String, Object>`.

- So we can choose the config form in the future (can be Kotlin dsl evaluated to a configuration, groovy, [HCL](https://github.com/hashicorp/hcl), etc...)
- Can change lib to parse (replace SnakeYaml with Jackson)
- Connect plugins that can just load ready-to-use programmatic configuration for third-party
- Even use GRPC to push config updates.

Also:
- Should make this sources list ordered
- Think on how it could be push-based instead of pull-based (for instance it could be git-repo with hooks/file watcher, so when change comes in, source can handle new objects and push new configuration to Jenkins)

**Comment by MadsNielsen** [comment]
@lanwen Thanks for the input, must admit i'm no expert in Java. :-) I think i'll let others implement this as my knowledge is lacking here and i'd be making too many errors which would mean it would just be rewritten in the next commit.

**Pull request closed without merging**